src/mpid/ch4/netmod/ofi/ofi_probe.h

```diff
@@ -111,14 +111,19 @@ static inline int MPIDI_NM_mpi_improbe(int source,
 
 #ifdef MPIDI_ENABLE_LEGACY_OFI
     if (!MPIDI_OFI_ENABLE_TAGGED) {
+        MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
```

Comment:
The `workq` config is orthogonal to `am-only` (or `legacy-ofi`) config, at least in master. Does `workq` works for `am-only` config in Intel's branch?

Reply:
Actually I with draw my previous comments. I guess the key here is `workq` lives in `ch4-layer`. At netmod layer, there are two branches: if there is a request passed down, we need to take the vci from the request. Otherwise, we create request from vci determined by netmod.

Currently, mulitvci in this PR only works with the latter, i.e. without workq. So I guess we need add an assertion that workq and multi-vci cannot be active simultaneously.

Separate work is needed to make multi-vci work with workq -- as well as am, rma -- so not in this PR.

Reply:
The request based solution that you described would help if we need to take a lock. However, for workq, I think we may not even need to take any lock at the netmod level.  I'm not sure if the current design supports this part.

Reply:
Current master does not support lockless mode. The netmod need ensure consistent locking -- for example, if send sends to a wrong vni, the recv end may never recv it. Therefore, it is difficult to let workq decide the vnis at ch4 layer. We may expose the netmod vni mapping function -- which adds the netmod api, probably shmmod api as well. I would like to explore the solution of using child request similar to persistent request for workq first. Anyway, workq will need separately work.

Reply:
> Where would we make that decision after this PR gets merged into the master? 

It is the same decision, but moved one layer down.

Actually I am not sure how current Intel branch would work. If you decide at ch4-layer to take lockless, that means you can't have active message and shmem since both will need lock?

Reply:
Currently, lockless is applicable only to ofi-netmod. Sure, deciding at ofi-netmod will suit better under the new design change. Do we expose the multi-threading models to ofi-netmod, so that it can make decisions based on what is used?

Reply:
Everything in the ch4-layer is exposed to the netmod, right?  Separately, since the lockless is ofi-only, we probably should change that config option as an ofi netmod option, for example, `--enable-ch4-ofi-lockless`?

Reply:
We should be able to use the things from ch4 at netmod. Yea, moving the lockless setup to the ofi level may be fine or fit well because the locking is moving there too.